onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/softmax_op_builder.cc

```diff
@@ -90,20 +94,67 @@ Status SoftMaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, cons
     y_scale = 1.f / 256;
   }
 
-  const auto& output = node_unit.Outputs()[0].node_arg.Name();
-  float beta = 1.f;
   InlinedVector<uint32_t> input_indices;
   input_indices.push_back(operand_indices.at(input));
+  const float beta = 1.f;
   ADD_SCALAR_OPERAND(model_builder, input_indices, beta);
 
-  if (android_feature_level > ANEURALNETWORKS_FEATURE_LEVEL_2) {
-    // you can only specify axis for android api level 29+
-    ADD_SCALAR_OPERAND(model_builder, input_indices, axis);
+  auto input_shape = shaper[input];
+  if (axis < 0) {
+    axis = static_cast<int32_t>(HandleNegativeAxis(axis, input_shape.size()));
   }
 
-  const OperandType output_operand_type(operand_types.at(input).type, shaper[output], y_scale, y_zero_point);
-  ORT_RETURN_IF_ERROR(model_builder.AddOperation(ANEURALNETWORKS_SOFTMAX, input_indices,
-                                                 {output}, {output_operand_type}));
+  // if opset < 13 we may need to manually coerce into 2D. we can skip this IFF it's already 2D and axis == 1.
+  // otherwise we need the coercion to create an input shape that works with axis == 1, which will work with any
+  // NNAPI version.
+  // e.g. if 2D and axis is 0 we coerce to shape {1 , dim0 + dim1}
+  //      if 4D and axis is 2 we coerce to shape {dim0 + dim1, dim2 + dim3}
+  if (node_unit.SinceVersion() < 13 && !(input_shape.size() == 2 && axis == 1)) {
+    // add Reshape to 2D
+    uint32_t dim0 = std::accumulate(input_shape.cbegin(), input_shape.cbegin() + axis, 1, std::multiplies());
+    uint32_t dim1 = std::accumulate(input_shape.cbegin() + axis, input_shape.cend(), 1, std::multiplies());
+    Shape input2d_shape{narrow<uint32_t>(dim0), narrow<uint32_t>(dim1)};
```

Comment:
shouldn't need `narrow` here as the argument is already uint32_t. however, the `std::accumulate` could overflow so maybe consider `SafeInt` there

Reply:
If the input shape has numbers so large it overflows I would assume we're running out of memory a lot earlier. :P